File: tools/aidkit/pkg/discovery/discovery.go
// Package discovery implements the AID file discovery protocol per spec §10.5.
package discovery

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/dan-strohschein/aidkit/pkg/parser"
)

const AidDocsDir = ".aidocs"
const ManifestFile = "manifest.aid"

// Result holds the discovery outcome.
type Result struct {
	AidDocsPath  string            // Absolute path to .aidocs/ directory
	ManifestPath string            // Path to manifest.aid (empty if not found)
	Manifest     *parser.AidFile   // Parsed manifest (nil if no manifest)
	AidFiles     []string          // All .aid files found
}

// Discover walks up from startDir looking for .aidocs/ per the spec protocol.
func Discover(startDir string) (*Result, error) {
	absStart, err := filepath.Abs(startDir)
	if err != nil {
		return nil, err
	}

	dir := absStart
	for {
		candidate := filepath.Join(dir, AidDocsDir)
		if info, err := os.Stat(candidate); err == nil && info.IsDir() {
			return inspectAidDocs(candidate)
		}

		parent := filepath.Dir(dir)
		if parent == dir {
			break // reached root
		}
		dir = parent
	}

	return nil, nil // no .aidocs/ found
}

func inspectAidDocs(aidDocsPath string) (*Result, error) {
	result := &Result{
		AidDocsPath: aidDocsPath,
	}

	// List .aid files
	entries, err := os.ReadDir(aidDocsPath)
	if err != nil {
		return nil, err
	}
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".aid") {
			result.AidFiles = append(result.AidFiles, e.Name())
		}
	}

	// Check for manifest
	manifestPath := filepath.Join(aidDocsPath, ManifestFile)
	if _, err := os.Stat(manifestPath); err == nil {
		result.ManifestPath = manifestPath
		manifest, _, err := parser.ParseFile(manifestPath)
		if err == nil {
			result.Manifest = manifest
		}
	}

	return result, nil
}

File: tools/aidkit-py/src/aidkit/discovery.py
"""AID file discovery protocol per spec section 10.5."""
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

from .model import AidFile
from .parser import parse_file

AIDOCS_DIR = ".aidocs"
MANIFEST_FILE = "manifest.aid"


@dataclass
class DiscoveryResult:
    """Holds the discovery outcome."""

    aidocs_path: str = ""
    manifest_path: str = ""
    manifest: AidFile | None = None
    aid_files: list[str] = field(default_factory=list)


def discover(start_dir: str | Path) -> DiscoveryResult | None:
    """Walk up from start_dir looking for .aidocs/ per the spec protocol."""
    current = Path(start_dir).resolve()

    while True:
        candidate = current / AIDOCS_DIR
        if candidate.is_dir():
            return _inspect_aidocs(candidate)

        parent = current.parent
        if parent == current:
            break
        current = parent

    return None


def _inspect_aidocs(aidocs_path: Path) -> DiscoveryResult:
    result = DiscoveryResult(aidocs_path=str(aidocs_path))

    # List .aid files
    for entry in sorted(aidocs_path.iterdir()):
        if entry.is_file() and entry.suffix == ".aid":
            result.aid_files.append(entry.name)

    # Check for manifest
    manifest_path = aidocs_path / MANIFEST_FILE
    if manifest_path.is_file():
        result.manifest_path = str(manifest_path)
        try:
            manifest, _ = parse_file(manifest_path)
            result.manifest = manifest
        except Exception:
            pass

    return result
